### PR TITLE
Enable JSPromise conversion for "anything Future-like"

### PR DESCRIPTION
**Pull request by axos88**
This allows chaining the implicit conversions such as going from a Q.Promise to a JSPromise, as in:
```scala
httpService.get[T](url)toJSPromise
```

### CODE REVIEW

**Comment by sjrd** [comment]
Thanks for your contribution, however I do not believe this is a good idea. Implicit chaining should be enabled with care, as it can quickly become difficult to track.

IMO it is better to explicitly enable such a chain with a user-defined implicit like this:
```scala
implicit def RichQPromise[A](p: Q.Promise[A]): js.JSConverters.JSRichFuture[A] =
  js.JSConverters.JSRichFutureNonThenable(p)
```
Also, your current implementation is not precise enough: since it is not possible to have a `Q.Promise[Thenable[A]]` (for the same reason that there cannot be a `Thenable[Thenable[A]]`, there is no need for the `Thenable`/`NonThenable` distinction.

File: library/src/main/scala/scala/scalajs/js/JSConverters.scala [hunk 1]
```diff
@@ -21,7 +21,7 @@ import scala.scalajs.runtime.genTraversableOnce2jsArray
 sealed abstract class JSConvertersLowPrioImplicits { this: JSConverters.type =>
 
   @inline
-  implicit def JSRichFutureNonThenable[A](f: Future[A]): JSRichFuture[A] =
+  implicit def JSRichFutureNonThenable[A, A1](f: A1)(implicit ev: A1 => Future[A]): JSRichFuture[A] =
     new JSRichFuture[A](f.asInstanceOf[Future[A | Thenable[A]]])
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSConverters.scala]: This completely by-passes `ev`. The cast applies without `ev` kicking in first. At least this should be `ev(f).asInstanceOf[...]`.

File: library/src/main/scala/scala/scalajs/js/JSConverters.scala [hunk 2]
```diff
@@ -63,7 +63,7 @@ object JSConverters extends JSConvertersLowPrioImplicits {
     new JSRichGenTraversableOnce(arr)
 
   @inline
-  implicit def JSRichFutureThenable[A](f: Future[Thenable[A]]): JSRichFuture[A] =
+  implicit def JSRichFutureThenable[A, A1](f: A1)(implicit ev: A1 => Future[Thenable[A]]): JSRichFuture[A] =
     new JSRichFuture[A](f.asInstanceOf[Future[A | Thenable[A]]])
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSConverters.scala]: Same problem with `ev` missing.

**Comment by axos88** [comment]
ev is missing, hold on

File: library/src/main/scala/scala/scalajs/js/JSConverters.scala (hunk 1 above)
Comment (axos88) [inline library/src/main/scala/scala/scalajs/js/JSConverters.scala]: Yeah, I just discovered this. Will update in a bit

**Comment by axos88** [comment]
I agree that implicit chaining should be taken with care however in this case I find it hard to find a case where it would casue surprises.

It is used only for extending "anything that can be converted to a Future" to be able to be converted to the JS equivalent JSPromise as well. That conversion however needs to be called explicitly, so converting it to a Future before converting it to the JSPromise must be the developer's intent.

**Comment by sjrd** [comment]
Hum sorry I may not have been clear enough. When I said only the case `NonThenable` was relevant, that applied only to `Q.Promise`. For general convertibles, both are necessary.

I am still not convinced this is a good idea. `Q.Promise` can have a better translation to `js.Promise` than this one.

Also, this change would require tests. And I'm afraid in any case or won't be binary compatible anyway.

**Comment by axos88** [comment]
Can you rephrase your last sentence?

**Comment by sjrd** [comment]
Oops sorry, phone auto correction. I edited my earlier comment.

**Comment by sjrd** [comment]
> `Q.Promise` can have a better translation to `js.Promise` than this one.

Actually even better than I thought. You can convert a `p: Q.Promise` to `js.Promise` natively with `js.Promise.resolve(p)`. So it's really better to define a conversion such as
```scala
implicit class QPromiseOps[A](val self: Q.Promise[A]) extends AnyVal {
  def toJSPromise: js.Promise[A] = js.Promise.resolve(self)
}
```

**Comment by mejine** [comment]
extends AnyVal {
  def toJSPromise: js.Promise[A] = js.Promise.resolve(self)

**Comment by sjrd** [comment]
@mejine Hum ... did you send your comment too soon? I don't understand it.

**Comment by mejine** [comment]
Hum sorry ??

**Comment by sjrd** [comment]
Well I did not understand your comment. It seems to be a copy-paste of part of my own comment. Basically it looks like your comment is missing some sentence or explanation or, well, something. Would you mind elaborating a bit on what you were trying to say?

**Comment by axos88** [comment]
Hmm. Have to check out wether this works with failed promises as well.
